plans/sep8.md

```diff
@@ -0,0 +1,187 @@
+- If regulated, notify user submitting request for transaction approval
+- Submit transaction to approval server, and depending on response:
+  - if success
+    - Display message if included in response
+    - Submit approved tx to the network
+  - if revised
+    - Display message
+    - Parse returned tx, display modifications to user
+    - Prompt user to resign tx
+    - Submit resigned tx to the network
+  - if pending
+    - Display timeout, message if included in response
+  - if action required
+    - Display message
+    - Redirect to action url
+  - if rejected
+    - "approval failed"
+    - Display error
+
+
+### Types
+
+```ts
+type getActionUrl = (args: GetActionArgs) => string;
+
+interface GetActionArgs {
+  request: ApprovalRequest;
+  response: ActionRequired;
+  callbackUrl: string;
+}
+
+class ApprovalProvider {
+  constructor(approvalServer, regulatedAssets) {}
+  approve: (args: ApprovalRequest) => Promise<ApprovalResponse>;
+  needsApproval: (args: Transaction) => boolean;
+  fetchActionInBrowser: ({
+    response: ActionRequired,
+    window: Window,
+  }) => Promise<ActionPromptStatus>;
+}
+
+interface ApprovalRequest {
+  tx: string;
+}
+
+interface ApprovalResponse {
+  status:
+    | "success"
+    | "revised"
+    | "pending"
+    | "action_required"
+    | "rejected";
+}
+
+interface TransactionApproved extends ApprovalResponse {
+  status: "success";
+  tx: string;
+  message?: string;
+}
+
+interface TransactionRevised extends ApprovalResponse {
+  status: "revised";
+  tx: string;
+  message: string;
+}
+
+interface PendingApproval extends ApprovalResponse {
+  status: "pending";
+  timeout: number;
+  message?: string;
+}
+
+interface ActionRequired extends ApprovalResponse {
+  status: "action_required";
+  message: string;
+  action_url: string;
+}
+
+interface ActionPromptStatus extends ApprovalResponse {
+  status: "success" | "rejected";
+  tx: string;
+}
+
+interface TransactionRejected extends ApprovalResponse {
+  status: "rejected";
+  error: string;
+}
+```
+
+### Usage
+
+```js
+import {
+  ApprovalProvider,
+  ApprovalResponseType,
+  getActionUrl
+} from "wallet-sdk";
+
+// approvalServerUrl and regulatedAssets fetched from stellar.toml
+// regulatedAssets is a map of `${asset.code}-${asset.issuer}` to 'stellar-base' Asset
+const approvalProvider = new ApprovalProvider(approvalServerUrl, regulatedAssets);
+
+// Parse transaction to check if involves regulated assets
+const needsApproval = approvalProvider.needsApproval(transaction);
+if (!needsApproval) {
+  // No approval needed so submit to the network
+  submitPayment(transaction);
+  return;
+}
+
+// Notify user asking for approval
+showUser(needsApproval);
+
+// Deconstruct to transaction envelope then submit to approval server
+// TODO: Maybe provider helper for deconstruct/reconstruct of tx <-> Transaction?
+const tx = transaction.toEnvelope().toXDR().toString('base64');
+const approvalResponse = await approvalProvider.approve({ tx });
```

Comment:
We'll probably want to pass the `Transaction` object to the approve function and let the library handle the implementation details.